pkg/cni/plugin.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func getEnvArgs(envArgsString string) (EnvArgs, error) {
+	if envArgsString != "" {
+		e := EnvArgs{}
+		err := types.LoadArgs(envArgsString, &e)
+		if err != nil {
+			return EnvArgs{}, err
+		}
+		return e, nil
+	}
+	return EnvArgs{}, nil
+}
+
+func getMTUByName(ifName string) (int, error) {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return 0, err
+	}
+	return link.Attrs().MTU, nil
+}
+
+func configureArp(macvtapConfig netlink.Link, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		// For sysctl, dots are replaced with forward slashes
+		name := strings.Replace(macvtapConfig.Attrs().Name, ".", "/", -1)
+
+		// TODO: duplicate following lines for ipv6 support, when it will be added in other places
+		ipv4SysctlValueName := fmt.Sprintf(IPv4InterfaceArpProxySysctlTemplate, name)
+		if _, err := sysctl.Sysctl(ipv4SysctlValueName, "1"); err != nil {
+			// remove the newly added link and ignore errors, because we already are in a failed state
+			_ = netlink.LinkDel(macvtapConfig)
+			return fmt.Errorf("failed to set proxy_arp on newly added interface %q: %v", macvtapConfig.Attrs().Name, err)
+		}
+		return nil
+	})
+	return err
+}
+
+func updateMacvtapIface(macvtapLink netlink.Link, macvtapIface *current.Interface, ifaceName string, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		err := ip.RenameLink(macvtapLink.Attrs().Name, ifaceName)
+		if err != nil {
+			_ = netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to rename macvlan to %q: %v", ifaceName, err)
+		}
+
+		updatedLink := macvtapLink
+		updatedLink.Attrs().Name = ifaceName
+		if err := netlink.LinkSetUp(updatedLink); err != nil {
+			return fmt.Errorf("failed to set macvtap iface up: %v", err)
+		}
+		// Re-fetch macvlan to get all properties/attributes
+		contMacvlan, err := netlink.LinkByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("failed to refetch macvlan %q: %v", ifaceName, err)
+		}
+		macvtapIface.Mac = contMacvlan.Attrs().HardwareAddr.String()
+		macvtapIface.Sandbox = netns.Path()
+
+		return nil
+	})
+	return err
+}
+
+func configureMacvtap(conf *NetConf, ifName string, netns ns.NetNS) (*current.Interface, error) {
+	iface, err := netlink.LinkByName(conf.DeviceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup device %q: %v", conf.DeviceID, err)
+	}
+	if err := netlink.LinkSetNsFd(iface, int(netns.Fd())); err != nil {
+		return nil, fmt.Errorf("failed to move iface %s to the netns %d because: %v", iface, netns.Fd(), err)
+	}
+	err = netns.Do(func(_ ns.NetNS) error {
+		if err := netlink.LinkSetMTU(iface, conf.MTU); err != nil {
+			return fmt.Errorf("failed to set the macvtap MTU for %s: %v", conf.DeviceID, err)
+		}
+		return nil
+	})
+	macvtap := &current.Interface{Name: ifName}
+	err = configureArp(iface, netns)
+	if err != nil {
+		return nil, err
+	}
+	err = updateMacvtapIface(iface, macvtap, ifName, netns)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I believe we should leave the netlink iface UP

Reply:
Explicitly done in the `CmdAdd`method.

Please resolve conversation if you agree.